### PR TITLE
First pass at basic matching with the FFI.

### LINKED ISSUE #44: Create draft design for message matching
We need the ability to match on messages, but there are open questions about what this API should look like. We'll need to work through the different capabilities in `pact_matching` that we ought to expose, how those can/should be used on the C side, and whether it can be split into parts for separate implementation.

### PR DESCRIPTION
**Pull request by alilleybrinker**
This commit introduces the first matching function of the FFI,
`match_message`, along with (and more importantly), a mechanism for
iterating over and inspecting mismatches which are returned by this
and other future matching methods.

Closes #44

### CODE REVIEW

File: rust/pact_matching_ffi/src/lib.rs
Comment (ewalters-mitre) [inline rust/pact_matching_ffi/src/lib.rs:90]: So the scope of mismatch types is wide enough that we need to fall back on a raw string instead of something like an enum?  Just clarifying.

Comment (alilleybrinker) [reply rust/pact_matching_ffi/src/lib.rs:90]: This is for consistency with the underlying implementation, and simplicity in the FFI API. Right now, `pact_matching` returns a string from `Mismatch::mismatch_type`, so what we're doing here is simply getting that string and converting it into a C-compatible representation. We _could_, in the FFI, define an enum which matches against the string from the Rust side, but it'd be brittle if the underlying `pact_matching` implementation decided to add new variants. I'm not opposed to it, but that's my perspective right now.

Comment (ewalters-mitre) [reply rust/pact_matching_ffi/src/lib.rs:90]: OK - I'm good with that then!

Comment (alilleybrinker) [reply rust/pact_matching_ffi/src/lib.rs:90]: Great! I'll wait for @cstepanian to give a review too before merging. Thanks!

Comment (cstepanian) [reply rust/pact_matching_ffi/src/lib.rs:90]: This looks fine to me --- this code is just exposing the available methods on the underlying `Mismatch` enum. I imagine this particular method should only be used for human-facing text output, and a client wanting the structure of the Mismatch could instead call `mismatch_to_json` and parse the JSON.
In the future, perhaps another interface could be added that gives the calling code a direct view of the enum, without having to go through JSON.

And while we're talking about this bit of code:
On line 89, it appears that "mismatch" is misspelled.

Comment (alilleybrinker) [reply rust/pact_matching_ffi/src/lib.rs:90]: Alright, I've fixed the typo. Thanks @cstepanian!

**Review by cstepanian — APPROVED** [review]

**Pull request merged**
